core/src/main/java/ch/cyberduck/core/DefaultHostPasswordStore.java

```diff
@@ -215,29 +217,56 @@ public void save(final Host bookmark) throws LocalAccessDeniedException {
         }
         if(credentials.isOAuthAuthentication()) {
             final String prefix = this.getOAuthPrefix(bookmark);
+            final String hostname = this.getOAuthHostname(bookmark);
+            final int port = getPortFromTokenUrl(bookmark);
+            final Scheme scheme = getSchemeFromTokenUrl(bookmark);
             if(StringUtils.isNotBlank(credentials.getOauth().getAccessToken())) {
-                this.addPassword(bookmark.getProtocol().getScheme(),
-                        bookmark.getPort(), this.getOAuthHostname(bookmark),
+                this.addPassword(scheme, port, hostname,
                         String.format("%s OAuth2 Access Token", prefix), credentials.getOauth().getAccessToken());
             }
             if(StringUtils.isNotBlank(credentials.getOauth().getRefreshToken())) {
-                this.addPassword(bookmark.getProtocol().getScheme(),
-                        bookmark.getPort(), this.getOAuthHostname(bookmark),
+                this.addPassword(scheme, port, hostname,
                         String.format("%s OAuth2 Refresh Token", prefix), credentials.getOauth().getRefreshToken());
             }
             // Save expiry
             if(credentials.getOauth().getExpiryInMilliseconds() != null) {
-                this.addPassword(this.getOAuthHostname(bookmark), String.format("%s OAuth2 Token Expiry", prefix),
+                this.addPassword(hostname, String.format("%s OAuth2 Token Expiry", prefix),
                         String.valueOf(credentials.getOauth().getExpiryInMilliseconds()));
             }
             if(StringUtils.isNotBlank(credentials.getOauth().getIdToken())) {
-                this.addPassword(bookmark.getProtocol().getScheme(),
-                        bookmark.getPort(), this.getOAuthHostname(bookmark),
+                this.addPassword(scheme,
+                        port, hostname,
                         String.format("%s OIDC Id Token", prefix), credentials.getOauth().getIdToken());
             }
         }
     }
 
+    private static int getPortFromTokenUrl(final Host bookmark) {
+        final String oAuthTokenUrl = bookmark.getProtocol().getOAuthTokenUrl();
+        int port = bookmark.getPort();
+        try {
+            URI url = URI.create(oAuthTokenUrl);
+            if(url.getPort() != -1) {
+                port = url.getPort();
+            }
+            else {
+                port = DefaultSchemePortResolver.INSTANCE.resolve(HttpHost.create(oAuthTokenUrl));
+            }
+        }
+        catch(UnsupportedSchemeException e) {
+            if(log.isWarnEnabled()) {
+                log.warn(String.format("Could not parse  %s, using protocol port instead.", oAuthTokenUrl), e);
+            }
+        }
+        return port;
+    }
+
+    private static Scheme getSchemeFromTokenUrl(final Host bookmark) {
+        final String oAuthTokenUrl = bookmark.getProtocol().getOAuthTokenUrl();
```

Comment:
Is `oAuthTokenUrl` always defined here? Default to `bookmark.getProtocol().getScheme()` if not?